CustomElementRegistry.prototype.initialize doesn't work on a single element

rniwa opened the pull request:
#### cbfaf8556c9e9733968bd24de9b9fe602f5ee974
<pre>
CustomElementRegistry.prototype.initialize doesn&apos;t work on a single element
<a href="https://bugs.webkit.org/show_bug.cgi?id=291723">https://bugs.webkit.org/show_bug.cgi?id=291723</a>

Reviewed by Anne van Kesteren.

Fix the bug in CustomElementRegistry::initialize which wasn&apos;t updating the registry of the root when it&apos;s not a shadow root.

This PR also changes so that Document now sets setUsesNullCustomElementRegistry() when it does not have a browsing context,
and fixes Document::createElement and friends to propagate the null registry status from Document to a newly created element.

Finally, this PR makes Document::customElementRegistryForBindings fallback to TreeScope::customElementRegistry() when no
LocalDOMWindow is associated with the document (i.e. registry is manually set by CustomElementRegistry::initialize).

This PR imports a WPT added in <a href="https://github.com/web-platform-tests/wpt/pull/52045">https://github.com/web-platform-tests/wpt/pull/52045</a> and updated in
<a href="https://github.com/web-platform-tests/wpt/pull/51976">https://github.com/web-platform-tests/wpt/pull/51976</a> and adds another WPT for more comprehensive testing of
CustomElementRegistry.prototype.initialize and customElementRegistry getter on Document and Element.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize.html: Added.
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::initialize):
* Source/WebCore/dom/CustomElementRegistry.h:
(WebCore::CustomElementRegistry::registryForNodeOrTreeScope):
* Source/WebCore/dom/Document.cpp:
(WebCore::m_syncData):
(WebCore::createUpgradeCandidateElement):
(WebCore::Document::createElementForBindings):
(WebCore::Document::createElement):
(WebCore::Document::customElementRegistryForBindings):
(WebCore::Document::createElementNS):
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::DocumentFragment):

Canonical link: <a href="https://commits.webkit.org/293968@main">https://commits.webkit.org/293968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/839af89d4069dafe3f708458b5fea9023109a626

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100434 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105571 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51022 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102475 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20393 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28560 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76471 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33524 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103441 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15624 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90721 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56827 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15440 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8724 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50397 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85356 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107925 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27552 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20218 "Found 1 new test failure: imported/w3c/web-platform-tests/IndexedDB/idbfactory-open-request-error.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85424 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27915 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86921 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84962 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29649 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7380 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21515 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16345 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27487 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32730 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27298 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30616 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28856 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk reviewed (approved):
Maybe you could also import https://github.com/web-platform-tests/wpt/pull/51976 while here and make sure those tests pass.

rniwa committed:
CustomElementRegistry.prototype.initialize doesn't work on a single element
https://bugs.webkit.org/show_bug.cgi?id=291723

Reviewed by Anne van Kesteren.

Fix the bug in CustomElementRegistry::initialize which wasn't updating the registry of the root when it's not a shadow root.

This PR also changes so that Document now sets setUsesNullCustomElementRegistry() when it does not have a browsing context,
and fixes Document::createElement and friends to propagate the null registry status from Document to a newly created element.

Finally, this PR makes Document::customElementRegistryForBindings fallback to TreeScope::customElementRegistry() when no
LocalDOMWindow is associated with the document (i.e. registry is manually set by CustomElementRegistry::initialize).

This PR imports a WPT added in https://github.com/web-platform-tests/wpt/pull/52045 and updated in
https://github.com/web-platform-tests/wpt/pull/51976 and adds another WPT for more comprehensive testing of
CustomElementRegistry.prototype.initialize and customElementRegistry getter on Document and Element.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize.html: Added.
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::initialize):
* Source/WebCore/dom/CustomElementRegistry.h:
(WebCore::CustomElementRegistry::registryForNodeOrTreeScope):
* Source/WebCore/dom/Document.cpp:
(WebCore::m_syncData):
(WebCore::createUpgradeCandidateElement):
(WebCore::Document::createElementForBindings):
(WebCore::Document::createElement):
(WebCore::Document::customElementRegistryForBindings):
(WebCore::Document::createElementNS):
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::DocumentFragment):

Canonical link: https://commits.webkit.org/293968@main
